Provide support for polygons and nongeospatial Records

jeancochrane opened the pull request:
# Overview

This PR proposes an updated data model providing support for polygons and nongeospatial Records.

## Notes

- This PR only focuses on updating the data model. Subsequent PRs will expose new functionality in the different Grout frontends, including:
    - Update frontend for Schema Editor to expose these new classes
    - Update JS API to expose new endpoints

## Testing instructions

Travis will run the existing unit test suite for all supported versions of Python and Django. To run the tests locally:

```bash
# Build containers and run migrations.
./scripts/update

# Run tests.
./scripts/test app
```

Note that this will install virtualenvs for every version of Python and Django in the support matrix. If you'd like to remove virtualenvs after you're done, make sure to run `sudo rm -Rf .tox` in the project directory.

Closes https://github.com/azavea/grout-2018-fellowship/issues/31, https://github.com/azavea/grout-2018-fellowship/issues/33, and https://github.com/azavea/grout-2018-fellowship/issues/43.

jeancochrane committed:
Sketch out model architecture for supporting non-geospatial records

jeancochrane committed:
Use abstract base models for flexible records

jeancochrane committed:
Remove reference to SchemaModel in tests

jeancochrane commented on grout/models.py:
```diff
@@ -24,14 +27,34 @@ class Meta(object):
 
 
-class SchemaModel(GroutModel):
+class RecordType(GroutModel):
```
I moved around some of the classes in this file to better approximate the hierarchical nature of the Grout data model (`RecordType -> RecordSchema -> Record`). This class, for instance, is now closer to the top.

jeancochrane commented on grout/models.py:
```diff
@@ -24,14 +27,34 @@ class Meta(object):
+        return schemas[0] if len(schemas) > 0 else None
+
+
+class RecordSchema(GroutModel):
```
Previously, `RecordSchema` had inherited from an abstract class `SchemaModel` that provided some JSON-validating methods. No other model was using `SchemaModel`, so I've merged it together with `RecordSchema` to simplify the data model a bit.

jeancochrane commented on grout/models.py:
```diff
@@ -52,43 +75,87 @@ def validate_schema(self, schema):
-    data = JSONField()
 
-    archived = models.BooleanField(default=False)
+class FlexibleRecord(AbstractFlexibleRecord):
```
`FlexibleRecord` and `TemporalFlexibleRecord` represent the non-geospatial types of Records introduced in this PR. While `Record` and `TemporalRecord` would probably be cleaner names, it seems important to me that we preserve the current `Record` class for backwards-compatibility.

jeancochrane commented on tests/test_schema_model.py:
```diff
@@ -2,19 +2,17 @@
+from grout.models import RecordSchema
 
 
 class SchemaModelTestCase(TestCase):
```
This test case needs to be adjusted to reflect the fact that the `SchemaModel` methods now reside directly on the `RecordSchema` model.

jeancochrane committed:
Add some failing tests for new API endpoints

jeancochrane committed:
Sketch out new geometry_type Record filter

jeancochrane committed:
Return different querysets depending on the Record query params

ddohler commented on grout/filters.py:
```diff
@@ -135,7 +145,44 @@ class Meta:
+
+
+class DateRangeFilterBackend(GenericFilterBackend):
+    # TODO: This class should be a method on the RecordFilter, not a backend. 
```
Is there any reason not to move this over now?

jeancochrane replied:
Nope! Planning to get on this today. I was just rushing out the door on Friday and this seemed like the easiest way to leave a note for myself to help get oriented on Monday.

ddohler commented on grout/models.py:
```diff
@@ -52,43 +75,138 @@ def validate_schema(self, schema):
+        ordering = ('-created',)
+
+
+class TemporalPointRecord(Record):
```
I think you might be able to get away with `TemporalPointRecord = Record`, but no matter what, I'd actually suggest doing it the other direction: implementing `TemporalPointRecord` explicitly and aliasing the backwards-compatible version. That way it's a smaller change to take it out later.

jeancochrane replied:
The main reason I aliased `TemporalPointRecord` to `Record` instead of doing it the other way around is that I didn't want [multi-table inheritance](https://docs.djangoproject.com/en/2.0/topics/db/models/#multi-table-inheritance) to mean that the only data stored in the `Record` table was a reference to the `TemporalPointRecord` table. My motivation here was somewhat lazy: I just didn't want to write a data migration to turn all `Record`s into `TemporalPointRecord`s and then implement the references on `Record`. It seemed like a more incremental change to add a new model that would reference the old `Record` model instead.

All that being said, it doesn't seem reasonable now that I write it out like that, so I'll do the slightly harder/more correct thing and implement the data model correctly. `Record = TemporalPointRecord` is an interesting approach, and might get around some of these migration issues anyway.

ddohler commented on tests/test_date_filter.py:
```diff
@@ -121,3 +130,34 @@ def test_range_parse_errors(self):
+        force_authenticate(min_is_none_req, self.user)
+        min_is_none_res = self.view(min_is_none_req).render()
+
+        import pdb
```
Stray `pdb` import here. The way I usually try to avoid this is to write `import pdb; pdb.set_trace()` instead of having it on multiple lines, which will cause most linters to flag the line.

jeancochrane replied:
Oops, good catch!

ddohler commented on grout/views.py:
```diff
@@ -47,13 +55,85 @@ def get_serializer_class(self):
-                       DjangoFilterBackend, DateRangeFilterBackend)
+    filter_backends = (InBBoxFilter, JsonBFilterBackend, DateRangeFilterBackend)
+
+    VALID_GEOTYPES = ['point', 'polygon', 'none']
```
I think that it would be good to include MultiPolygons as well; there are plenty of real-world geometries that can't be expressed using polygons alone.

jeancochrane replied:
True -- and as long as we're using `GeometryField`, I agree that LineStrings are a reasonable addition, too.

ddohler commented on grout/views.py:
```diff
@@ -47,13 +55,85 @@ def get_serializer_class(self):
+        '''
+        Check for the `geometry_type` and `occurred_min`/`occurred_max`
+        query parameters. If they exist, use them to generate a queryset for
+        the appropriate class of Record.
```
My experience in the past has been that endpoints that change their behavior based on specific combinations of query parameters end up being pretty confusing and difficult to debug. One idea I had that might simplify things is to move the datetime and geometry type information up to the `RecordType` model. Then, this view could grab the relevant `RecordType` and use the flags on that model to figure out which type of queryset to use.

The downsides that I see to this are:
- It would make the `record_type` parameter mandatory. In practice this is already the case since we don't have a use case yet that I know of where querying for records from multiple different `record_types` with the same set of other parameters makes sense, so I think we always include it anyway. We might even think about enshrining the requirement in the API by making an endpoint at `/api/record_types/<UUID>/records/` that does the same thing.
- It would add a second database query to every `Record` request. My hunch is that this wouldn't be a big deal because all of our current use cases have relatively few (single digits) `RecordTypes` but lots and lots of `Records`, so my expectation would be that the `RecordType`s would be kept fairly warm in both the Django and Postgres caches. Still, it's something to keep in mind.
- It would prevent having `Record`s with varying geometry types within the same RecordType. I think this is likely to be a valuable restriction in that it'll probably simplify querying logic, and doesn't seem very likely to be something that anyone would want to do anyway.

How does that sound to you? Are there any drawbacks to this approach that I'm missing?

jeancochrane replied:
That's an interesting idea! For clarity, when you suggest moving geometry type and datetime information up to `RecordType`, are you thinking something like this:

```
RecordType(GroutModel):
    geometry_type = models.ChoiceField()
    temporal = models.BooleanField(default=True)
```

Points 1 and 2 above seem reasonable to me, although I'm on the fence about 3. I think the fliers app provides an interesting example edge case: for this use-case, polygons are attractive because sometimes you don't know the exact address of an event (or the event organizers don't want it to be made public), so the archivists would want to store a more general boundary instead of a point. In this case, it feels most intuitive for the two cases (a specific address point vs. a general boundary) to be the same type of Record, since they're attempting to catalog the same type of entities in the world. But I could imagine an implementation that separates them; you'd just have to have two record types, e.g. `Event` and `BoundaryEvent`, and make sure to query for both when appropriate. Plus, from a code perspective it might end up being more tedious to receive `Record`s with mixed geometries from the same API query, since then you have to do an additional filter to figure out how to display them.

All that being said, at this point I think the benefits of your proposal outweigh the costs. I can't think of any kind of query functionality that is made impossible by this change -- it just seems like it makes certain operations (querying for all `Record`s regardless of `RecordType`, querying for multiple geometry types) more verbose, but there's still uncertainty in whether those operations would be attractive to most users anyway.

ddohler replied:
That's an interesting use-case! So the idea is either to a) convey uncertainty about a location or b) to obscure the true location? I agree that storing these types of records as mixed geometries would be a bit of a pain to handle on the frontend. 

I wonder if there might be other ways to get at the same functionality that would keep filtering easy. For example, would it work to add a checkbox to the data entry form to obscure the true location? If that box was checked, the frontend could jitter the entered point by a certain amount, say 100 meters, before POSTing to the API. For uncertain location, perhaps the schema could have a toggle or numeric entry for uncertainty that would cause the frontend to draw a bubble rather than a point for that record if it was checked?

I think that handling uncertainty and obscuring true locations (potentially with the option to display it to administrators) are a common enough need that it might be helpful to add functionality to help with this to Grout. I've definitely encountered apps that have to grapple with one or more of these things before, but I haven't put much thought into how Grout might make it easier to handle these types of issues specifically.

ddohler commented on grout/models.py:
```diff
@@ -52,43 +75,138 @@ def validate_schema(self, schema):
+        POINT = 'point'
+        CHOICES = ((POINT, 'Point'),)
+
     geom = models.PointField(srid=settings.GROUT['SRID'])
```
I had a thought about this that might cut down on the number of subclasses and mixin we need to create: instead of specifying each geometry type individually, we could use a `GeometryField` instead to cover all geometry types with one field. This would immediately provide support for all non-collection [geometry types](https://docs.djangoproject.com/en/2.0/ref/contrib/gis/model-api/#module-django.contrib.gis.db.models) rather than requiring extra work to add `MultiPolygon`s, `LineString`s, etc. The major drawback I see is that we might lose some database-level validation, but I think we could simulate something similar by using [`Cast`](https://docs.djangoproject.com/en/2.0/ref/contrib/gis/model-api/#geography-type) when necessary (though I admittedly haven't tried using Cast for that purpose before). How does that sound?

jeancochrane replied:
I like this idea a lot! The class inheritance that's currently implemented in this PR makes the fields nicely composable, but I think it's reached a point where there are enough base classes and mixins now that it would be confusing to read if I didn't already understand the data model.

In terms of data validation, were you thinking of continuing to use a `geometry_type` field (or equivalent) in order to determine how the `Cast` should behave? That sounds like a promising idea to me, and it suggests that we could much more easily expand geometry support in the future without having to introduce millions of mixins. I haven't used `GeometryField` or `Cast` before and the GeoDjango docs are extremely sparse, so I have a suspicion that there may be a serious problem lurking there, but it seems reasonable to spend some time experimenting with that approach by setting up tests, sketching out the API, and seeing if it works.

ddohler replied:
Yep, that's exactly what I was thinking for validation. I haven't used this particular combination of features either, so it's definitely possible that there are some problems hiding somewhere, but yeah, I agree it seems worth a shot!

ddohler reviewed (changes requested):
It's awesome to see this started! I found a few minor things that need to be cleaned up, and I left some thoughts about larger structural changes that might prove helpful--let me know what you think!

jeancochrane committed:
Sketch out tests for new syntax for handling geometries

jeancochrane committed:
Remove provisional tests for scrapped API syntax

jeancochrane committed:
Strip out old changes for proposed API syntax

jeancochrane committed:
Add geometry_type validation and get tests passing

jeancochrane committed:
Implement filters and validation for nontemporal RecordTypes

jeancochrane committed:
Make sure DateFilter uses the occurred_to param as an upper bound

jeancochrane committed:
Move logic from DateRangeFilterBackend to method filters on the RecordFilter

jeancochrane committed:
Reorganize tests to split up monolithic RecordViewTestCase

jeancochrane committed:
Clean up diff for review

jeancochrane commented:
Alright, this is ready for another look @ddohler!

A couple notes:

- Differentiating geometries with `GeometryField` was surprisingly easy, and I didn't even have to do any fancy `Cast`ing! the Django REST Framework GIS package comes with some really nice built-in serializers for `GeometryField`s that read the `geom_type` of the geometry and spit out the appropriate GeoJSON.
- I didn't wind up making `record_type` a required field, since it turned out to be unnecessary for this pattern. Moving away from the class inheritance pattern and instead shifting the geospatial/temporal data modelling burden to `RecordType` simplified things enormously, with the great benefit of allowing us to tie the `api/record/` filters to the `Record` model again. If you think there's something to be gained by enforcing `record_type` as a requirement I'd be happy to talk it over, but it seems to me like we might as well leave the option open.
- I added a bunch of test cases to cover basic CRUD/filtering operations for all of the geotype/temporal combinations I could think of; let me know if I missed anything obvious.

ddohler commented on grout/filters.py:
```diff
@@ -68,19 +80,48 @@ def filter_polygon_id(self, queryset, field_name, poly_uuid):
+        if not min_date.tzinfo:
+            raise QueryParameterException('occurred_min', DATETIME_FORMAT_ERROR)
+        else:
+            return queryset.filter(occurred_from__gte=min_date)\
```
You could avoid the backslash here by passing the second condition as a second keyword argument to the first `filter` call, i.e. `queryset.filter(occurred_from__gte=min_date, occurred_to__gte=min_date)`.

I also wanted to call out that I suspect different users may have different expectations about what they want this filter to do; this will exclude any records where the `occurred_from` OR `occurred_to` fall earlier than the `min_date`, but I could easily imagine situations (such as using `occurred_from` to express uncertainty) where that might not be desirable. I don't think it's anything we need to worry about right now, but it'd be good to make sure the exact filtering behavior is clearly documented somewhere.

jeancochrane replied:
Your point about the behavior of the `occurred_min`/`occurred_max` filters is a really good one. Thinking about it more closely, it seems like the question the filters are currently answering is "Is the date range entirely contained within the filter range?" Perhaps a more intuitive question for the filter to answer would be "Does any part of the date range overlap with the filter range?"

Formally, this would amount to a change from

```
(from >= min AND to >= min) AND (from <= max AND to <= max)
```

to something like

```
from <= max AND to >= min
```

Does this logic seem sound? I think you're right that there will always be applications where one filter is preferable to the other, so this will definitely need to be well-documented. But I think as long as we're defining the default functionality, the "overlap" filter is probably a more sensible default than the "containment" filter, particularly since it's more forgiving of the end user.

ddohler replied:
Hmm, assuming that `to >= from` (side thought: we should probably add validation for that if it doesn't exist already), I think that the first and second version are functionally identical. Perhaps you meant `from <= max OR to >= min`? I think that would match the written description you gave (which I agree with). Regardless, yeah, I think either way is reasonable, but the more flexible one will probably result in the fewest surprises for new users so that seems like a pretty good default.

jeancochrane replied:
Agreed that we should validate `to >= from`! Seems simple enough to add in this PR.

OK, so this is definitely not critical to the discussion at hand, but in case it's interesting to you: I drew some pictures to help me think through it and I still think we need both `from <= max` and `to >= min` in order for the condition to hold. As an example, if `from < max` while `to < min` (which would satisfy `from <= max OR to >= min` but not `from <= max AND to >= min`), then a range like this one passes even though there's no overlap:

```
from               to
 |-----------------|
                                x          x
                               min        max

```

I don't think the first two conditions are equal either. Here's an example of a valid "overlap" case where `from <= max AND to >= min` passes, but the original "containment" condition (`(from >= min AND to >= min) AND (from <= max AND to <= max)`) doesn't, because `from < min)`:

```
from               to
 |-----------------|
                x          x
               min        max

```

This overlap also doesn't pass the "containment" condition, because `from < min` and `to > max`, but it does pass both conditions we suggested (`from <= min AND/OR to >= max`):

```
from               to
 |-----------------|
     x          x
    min        max
```

This wouldn't be the first time that range comparisons have got my brain tangled up, so it's entirely possible I've missed something important here!

ddohler replied:
You're exactly right! I've run through this exact problem (with pictures) a few times before so I foolishly thought I could get away with not drawing pictures this time, but nope! :+1:

ddohler commented on grout/models.py:
```diff
@@ -53,42 +96,73 @@ def validate_schema(self, schema):
+        if errors.keys():
+            # Raise a DRF ValidationError instead of a Django Core validation
+            # error, since this exception needs to get handled by the serializer.
+            raise serializers.ValidationError(errors)
```
This is a bit of an odd question because I thought that this functionality already existed, but... shouldn't saving a Record cause the Record's `data` field to be validated against `self.schema.schema`? Or is that happening somewhere else that I've forgotten about?

jeancochrane replied:
That's a very good question! I assumed that functionality existed too, but I took a closer look at the code and it seems to me like it doesn't eixst.

The `RecordSchema` model has a [`validate_json` method](https://github.com/azavea/grout/blob/c217cda8614c8d20f9b537c8759433efa9fad516/grout/models.py#L36-L43) that seems primed to perform this kind of validation:

```python
def validate_json(self, json_dict):
    """Validates a JSON-like dictionary against this object's schema
    :param json_dict: Python dict representing json to be validated against self.schema
    :return: None if validation succeeds; jsonschema.exceptions.ValidationError if failure
             (or jsonschema.exceptions.SchemaError if the schema is invalid)
    """
    return jsonschema.validate(json_dict, self.schema)
```

However, a quick grep shows that the method isn't called anywhere in the module:

```
$ grep -R -F "validate_json" grout
grout/validators.py:        def validate_json_schema(value):
grout/serializer_fields.py: from grout.validators import validate_json_schema
grout/serializer_fields.py: validators = [validate_json_schema]
grout/models.py:            def validate_json(self, json_dict):
```

As the grep shows, there is a `validate_json_schema` validator [defined in `validators.py`](https://github.com/azavea/grout/blob/c217cda8614c8d20f9b537c8759433efa9fad516/grout/validators.py#L7-L12) that gets used as a Django REST Framework validator for all incoming Records, but that method only validates that the incoming data is valid JSONSchema, not that it matches the Record's schema specifically:

```python
def validate_json_schema(value):
    """Wrap jsonschema validation with serializers.ValidationError"""
    try:
        jsonschema.Draft4Validator.check_schema(value)
    except SchemaError as e:
        raise ValidationError('Invalid schema: {}'.format(e.message))
```

My guess is that this validation all happens in the schema editor, and the logic isn't wired into Grout core. Seems like something we should add!

ddohler replied:
Thanks for digging into this! I think we should definitely add this, I'm kinda surprised it doesn't already exist but I must have been thinking of `validate_json_schema`. That could be a future task though.

jeancochrane replied:
Opened up https://github.com/azavea/grout-2018-fellowship/issues/44 to keep track of this task.

ddohler commented on tests/test_date_filter.py:
```diff
@@ -49,50 +49,94 @@ def setUp(self):
+            schema=self.nontemporal_schema,
+            geom='POINT (0 0)',
+            data={}
+        )
```
[factory-boy](http://factoryboy.readthedocs.io/en/latest/) can be super helpful with these kinds of chained model constructions, but it may not be worth the effort at this stage compared to some other things we want to do.

jeancochrane replied:
Oooh, nice! Will definitely consider using this if the fixture setup starts getting more complicated.

ddohler commented on grout/filters.py:
```diff
@@ -30,9 +31,20 @@
 
+    # Custom filter definitions.
     record_type = django_filters.Filter(field_name='record_type', method='filter_record_type')
     polygon = django_filters.Filter(field_name='polygon', method='filter_polygon')
```
One fairly nifty side effect of all this is that we can do much more interesting filtering than point-in-polygon without any changes. The filter constructs a `GEOSGeometry` from the incoming GeoJSON and then does an intersection, so it should be possible (for example) to filter polygons by a point (e.g. for address-in-jurisdiction type queries), or a linestring (e.g. all the parcels that a creek passes through).

The downside is that I think that means that this query parameter is now probably a misnomer, but I don't think you necessarily need to worry about that refactor right now unless you want to.

jeancochrane replied:
I love getting features for free! I think an easy way to help clear this up while doing the least mount of refactoring would be to add a new filter as an alias to the existing `polygon` filter, and to be sure to document the behavior using the new name. What do you think would be a good name for the filter? Is `intersects` or `intersection` too vague? Maybe `geointersects`?

ddohler replied:
I like that approach and I like all those names. I'll toss in a few more ideas, pick whichever you like best:
- `intersects_with`
- `intersects_geom`
- `geom_intersection`

ddohler reviewed (approved):
This looks great! I'm glad that the RecordType restructuring helped clean things up, and I agree -- I don't think we need to make the `record_type` parameter mandatory. I had a few comments but nothing that I think will change the core structure any further.

jeancochrane committed:
Change date range filter to use 'overlap' test instead of 'containment'

jeancochrane committed:
Check Record filters to make sure that occurred_min < occurred_max

jeancochrane committed:
Alias 'geom_intersects' query param to existing 'polygon' param

jeancochrane committed:
Update CHANGELOG

jeancochrane commented:
Alright, added in some new features and tests according to your comments @ddohler, including:

- Modified date range filter to use an "overlap" approach instead of "containment" (plus added more verbose tests)
- Added a validation check to make sure `occurred_min < occurred_max` when both query params are present
- Alias `geom_intersects` to the existing `polygon` query param

I also updated the `CHANGELOG` with all the changes made in this branch, since I noticed I had forgotten to do that earlier!

Would you like to take another look, or is this good to go?

ddohler commented:
I was actually referring to whether we validate that `occurred_from <= occurred_to`, but validating the filter parameters is good too! But I think we should also make sure that users can't create records that end before they start. Good to go after that. 👍

jeancochrane committed:
Validate that 'occurred_from' <= 'occurred_to' when creating a new Record

jeancochrane commented:
Ah, great point! Unfortunately not the first time I've gotten the model fields and the filters mixed up :sweat_smile: Just added a simple check and a unit test for cases where `occurred_from > occurred_to`.
